File: endpoint.py
import io
import json
from http.server import BaseHTTPRequestHandler, HTTPServer
from urlparse import urlparse
import unicodedata

F_NAME = 'name'
F_LONGITUDE = 'longitude'
F_LATITUDE = 'latitude'
F_COUNTRY = 'country'
F_REGION = 'region'
F_SCORE = 'score'

class CVTrie():

    class CVTrieNode:
        def __init__(self, char, cid):
            self.char = char
            self.ids = set()
            self.ids.add(cid)
            self.nodes = {}

        def __repr__(self):
            return '%s %s %s' % (self.char, self.ids, self.nodes)

    def __init__(self):
        print('CVTrie::init')
        self.cities = {}
        self.nodes = {}

    def __repr__(self):
        return '%s' % (self.nodes.keys())

    def find(self, prefix, latitude, longitude):
        print('CVTrie::find')
        result = []

        # walk the trie until exhaustion of the prefix
        node = self
        for c in prefix.lower():
            # skip non [a-z] characters
            if ord(c) < 97 or ord(c) > 122:
                continue
            node = node.nodes.get(c, None)
            if node is None:
                # not found
                return result

        # process the CIDs
        for cid in node.ids:
            city = self.cities[cid]
            resultNode = {
                F_NAME: '%s, %s, %s' % (city[F_NAME], city[F_REGION], city[F_COUNTRY]),
                F_LATITUDE: str(city[F_LATITUDE]),
                F_LONGITUDE: str(city[F_LONGITUDE])
            }
            resultNode[F_SCORE] = self.score(cid, prefix, latitude, longitude)
            result.append(resultNode)

        # sort in reverse order as per spec
        result.sort(cmp=lambda a, b: 1 if a[F_SCORE] < b[F_SCORE] else -1)

        return result

    def score(self, cid, query, latitude, longitude):
        print('CVTrie::score')
        city = self.cities[cid]
        print('CVTrie::score parameters', city, latitude, longitude)

        # name 40%
        divisor = max(len(query), len(city)) / min(len(query), len(city))
        prefField = min(40.0 / divisor, 40.0)

        # latitude 30%
        divisor = abs(latitude - city[F_LATITUDE])
        latField = min(30.0 / divisor, 30.0)

        # longitude 30%
        divisor = abs(longitude - city[F_LONGITUDE])
        longField = min(30.0 / divisor, 30.0)

        print('CVTrie::score components', prefField, latField, longField)
        return round((prefField + latField + longField) / 100.0, 1)

    def add(self, cid, name, asciiName, latitude, longitude, country, region):
        self.cities[cid] = {
            F_NAME: name,
            F_LONGITUDE: longitude,
            F_LATITUDE: latitude,
            F_COUNTRY: country,
            F_REGION: region
        }

        node = self
        indexable = list('abcdefghijklmnopqrstuvwxyz')
        extended = list('-\'., ()1')
        for c in asciiName.lower():
            if c not in indexable:
                if c not in extended:
                    print('Unknown character [%s] in [%s]' % (c, name))
                continue
            child = node.nodes.get(c, None)
            if child is None:
                child = self.CVTrieNode(c, cid)
                node.nodes[c] = child
            else:
                child.ids.add(cid)
            node = child

    def load(self, path):
        regionLookup = {
            '01': 'AB',
            '02': 'BC',
            '03': 'MB',
            '04': 'NB',
            '05': 'NL',
            '07': 'NS',
            '08': 'ON',
            '09': 'PE',
            '10': 'QC',
            '11': 'SK',
            '12': 'YT',
            '13': 'NT',
            '14': 'NU'
        }

        try:
            legend = None
            with io.open(path, mode='r', encoding='utf-8') as lines:
                for line in lines:
                    fields = line.strip().split('\t')
                    if legend is None:
                        # not used, serves as an init marker here
                        legend = fields
                    else:
                        # fields: 0: id | 1: name | 2: ascii | 3: altname | 4: lat | 5: long | 8: country | 10: region
                        if fields[8] == 'US':
                            region = fields[10]
                        else:
                            region = regionLookup.get(fields[10], '')
                            if not region:
                                print('Failed region code lookup:', fields)
                                continue
                        if fields[8] == 'US':
                            country = 'USA'
                        else:
                            country = 'Canada'
                        self.add(fields[0], fields[1], fields[2], float(fields[4]), float(fields[5]), country, region)
        except ValueError:
            print('Could not load data')

# http request handler implementation
class CVRequestHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        print('CVRequestHandler::do_GET url:[%s]' % self.path)
        query = urlparse(self.path).query
        params = {}
        if query:
            paramKVs = query.split('&')
            if paramKVs:
                params = dict(paramKV.split('=') for paramKV in paramKVs)

        # convert non-ascii characters
        prefix = params.get('q', None)
        prefix = bytes(prefix).decode('utf-8', 'strict')
        prefix = unicodedata.normalize('NFD', prefix).encode('ascii', 'ignore').decode('utf-8')

        # if not provided, use a minimal latitute or longitude
        try:
            latitude = float(params.get(F_LATITUDE, '0.00001'))
        except ValueError:
            latitude = 0.00001

        try:
            longitude = float(params.get(F_LONGITUDE, '0.00001'))
        except ValueError:
            longitude = 0.00001

        results = vtrie.find(prefix, latitude, longitude)

        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.end_headers()
        response = {
            'suggestions': results
        }
        self.wfile.write(json.dumps(response).encode('utf-8'))
        self.wfile.close()

# wrapper for the http listener
class CVService:
    def start(self, handlerClass=CVRequestHandler):
        print('CVSerice::start')
        server_address = ('', 8080)
        httpd = HTTPServer(server_address, handlerClass)
        saddr = httpd.socket.getsockname()
        print('Serving Endpoint at [%s:%i]' % (saddr[0], saddr[1]))
        httpd.serve_forever()

if __name__ == '__main__':
    vtrie = CVTrie()
    vtrie.load('cities_canada-usa.tsv')
    svc = CVService()
    svc.start()

File: endpoint.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
	"unicode"

	"golang.org/x/text/runes"
	"golang.org/x/text/transform"
	"golang.org/x/text/unicode/norm"
)

// City data
type City struct {
	name      string
	longitude float64
	latitude  float64
	country   string
	region    string
}

// QueryResultEntry data
type QueryResultEntry struct {
	Name      string  `json:"name"`
	Latitude  string  `json:"latitude"`
	Longitude string  `json:"longitude"`
	Score     float64 `json:"score"`
}

// QueryResult JSON input
type QueryResult struct {
	Suggestions []*QueryResultEntry `json:"suggestions"`
}

func newQueryResult() *QueryResult {
	return &QueryResult{Suggestions: make([]*QueryResultEntry, 0)}
}

func (result *QueryResult) add(name string, latitude float64, longitude float64, score float64) {
	result.Suggestions = append(result.Suggestions, &QueryResultEntry{
		Name:      name,
		Latitude:  strconv.FormatFloat(latitude, 'f', 5, 64),
		Longitude: strconv.FormatFloat(longitude, 'f', 5, 64),
		Score:     score,
	})
}

// NewCity helper
func NewCity(name string, latitude float64, longitude float64, country string, region string) *City {
	return &City{
		name:      name,
		longitude: longitude,
		latitude:  latitude,
		country:   country,
		region:    region,
	}
}

// TrieNodeMap typedef for map of trie nodes
type TrieNodeMap map[rune]*TrieNode

// TrieNode is the node definition for the trie
type TrieNode struct {
	char  rune
	ids   []int64
	nodes TrieNodeMap
}

// NewTrieNode helper
func NewTrieNode(char rune, cid int64) *TrieNode {
	return &TrieNode{
		char:  char,
		ids:   []int64{cid},
		nodes: make(TrieNodeMap),
	}
}

func (trieNode *TrieNode) addCid(cid int64) {
	trieNode.ids = append(trieNode.ids, cid)
}

// Trie is the main data structure
type Trie struct {
	cities map[int64]*City
	nodes  TrieNodeMap
}

// NewTrie helper
func NewTrie() *Trie {
	return &Trie{
		cities: make(map[int64]*City),
		nodes:  make(map[rune]*TrieNode),
	}
}

func (trie *Trie) find(prefix string, latitude float64, longitude float64) *QueryResult {
	result := newQueryResult()

	var node *TrieNode
	// walk the trie until exhaustion of the prefix
	for _, c := range strings.ToLower(prefix) {
		// skip [a-z] query characters
		if c < 97 || c > 122 {
			continue
		}

		if node == nil {
			// assign an initial node
			node = trie.nodes[c]
		} else {
			node = node.nodes[c]
		}
		if node == nil {
			// not found
			return result
		}
	}

	// failed to find trie value
	if node == nil {
		println("Failed to find trie entry for", prefix, latitude, longitude)
		return result
	}

	// process the CIDs
	for _, cid := range node.ids {
		city := trie.cities[cid]
		if city != nil {
			fullName := fmt.Sprintf("%s, %s, %s", city.name, city.region, city.country)
			score := trie.score(cid, prefix, latitude, longitude)
			result.add(fullName, city.latitude, city.longitude, score)
		}
	}

	// sort the results in descending order of score
	sort.Slice(result.Suggestions, func(i, j int) bool {
		return result.Suggestions[i].Score > result.Suggestions[j].Score
	})

	return result
}

func (trie *Trie) score(cid int64, query string, latitude float64, longitude float64) float64 {
	city := trie.cities[cid]

	divisor := math.Max(float64(len(query)), float64(len(city.name))) / math.Min(float64(len(query)), float64(len(city.name)))
	prefField := math.Min(40.0/divisor, 40.0)

	divisor = math.Abs(latitude - city.latitude)
	latField := math.Min(30.0/divisor, 30.0)

	divisor = math.Abs(longitude - city.longitude)
	longField := math.Min(30.0/divisor, 30.0)

	return math.Round((prefField+latField+longField)/100.0*10) / 10
}

var extendedLookup = map[string]bool{"-": true, "'": true, ".": true, ",": true, " ": true, "(": true, ")": true, "1": true}

func (trie *Trie) add(cid int64, name string, asciiName string, latitude float64, longitude float64, country string, region string) bool {
	city := NewCity(name, latitude, longitude, country, region)
	trie.cities[cid] = city

	var node *TrieNode
	var child *TrieNode

	for i, c := range strings.ToLower(asciiName) {
		// is c within [a-z]
		if c < 97 || c > 122 {
			if extendedLookup[string(c)] == false {
				println("Unknown character [%s] at [%i] in [%s]", c, i, name)
			}
			continue
		}
		if node == nil {
			child = trie.nodes[c]
		} else {
			child = node.nodes[c]
		}
		if child == nil {
			child = NewTrieNode(c, cid)
			if node == nil {
				trie.nodes[c] = child
			} else {
				node.nodes[c] = child
			}
		} else {
			child.addCid(cid)
		}
		node = child
	}

	return true
}

var regionLookup = map[string]string{
	"01": "AB",
	"02": "BC",
	"03": "MB",
	"04": "NB",
	"05": "NL",
	"07": "NS",
	"08": "ON",
	"09": "PE",
	"10": "QC",
	"11": "SK",
	"12": "YT",
	"13": "NT",
	"14": "NU",
}

// Load the tsv file and parse into a trie
func (trie *Trie) load(path string) bool {
	// file, err := os.Open(path)
	// defer file.Close()
	// if err != nil {
	// 	log.Fatal(err)
	// 	return false
	// }
	// scanner := bufio.NewScanner(file)
	// for scanner.Scan() {
	// 	fmt.Println(scanner.Text())
	// }

	timer := time.Now()
	bytes, err := ioutil.ReadFile(path)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// fields: 0: id | 1: name | 2: ascii | 3: altname | 4: lat | 5: long | 8: country | 10: region
	var lines = strings.Split(string(bytes), "\n")
	for i, line := range lines {
		fields := strings.Split(line, "\t")
		if len(fields) < 10 || i == 0 {
			println("Skipping line:", i)
			continue
		}
		var region string
		if fields[8] == "US" {
			region = fields[10]
		} else {
			region = regionLookup[fields[10]]
			if region == "" {
				println("Failed region code lookup:", fields)
				continue
			}
		}
		var country string
		if fields[8] == "US" {
			country = "USA"
		} else {
			country = "Canada"
		}
		cid, err0 := strconv.ParseInt(fields[0], 10, 64)
		latitude, err1 := strconv.ParseFloat(fields[4], 64)
		longitude, err2 := strconv.ParseFloat(fields[5], 64)
		if err0 == nil {
			if err1 != nil {
				latitude = 0.00001
			}
			if err2 != nil {
				longitude = 0.00001
			}
			trie.add(cid, fields[1], fields[2], latitude, longitude, country, region)
		}
	}

	println("Trie built in", time.Since(timer)/time.Millisecond, "ms")

	return true
}

func (trie *Trie) defaultHandler(resp http.ResponseWriter, req *http.Request) {
	req.ParseForm()
	q := req.FormValue("q")

	// strip extended characters
	t := transform.Chain(norm.NFD, runes.Remove(runes.In(unicode.Mn)), norm.NFC)
	prefix, _, _ := transform.String(t, q)

	latitude, errLat := strconv.ParseFloat(req.FormValue("latitude"), 64)
	longitude, errLong := strconv.ParseFloat(req.FormValue("longitude"), 64)

	if errLat != nil {
		latitude = 0.00001
	}
	if errLong != nil {
		longitude = 0.00001
	}

	timer := time.Now()
	result := trie.find(prefix, latitude, longitude)
	println("Search for", prefix, "completed in", time.Since(timer)/time.Microsecond, "μs")

	b, err := json.Marshal(result)
	if err != nil {
		println("JSON serialization error", err)
	}

	resp.Header().Set("Content-Type", "application/json; charset=utf-8")
	resp.Write([]byte(b))
}

func main() {
	trie := NewTrie()
	trie.load("cities_canada-usa.tsv")
	http.HandleFunc("/", trie.defaultHandler)
	println("Serving on port 8080")
	http.ListenAndServe(":8080", nil)
}
